src/main/scala/com/foursquare/common/thrift/bson/TBSONBinaryProtocol.scala

```diff
@@ -0,0 +1,280 @@
+    case BSON.ARRAY => TType.LIST
+    case BSON.BINARY => TType.STRING
+    case BSON.UNDEFINED => TType.VOID
+    case BSON.OID => TType.STRING
+    case BSON.BOOLEAN => TType.BOOL
+    case BSON.DATE => TType.I64
+    case BSON.NULL => TType.VOID
+    case BSON.REGEX => TType.STRING
+    case BSON.REF => TType.STRING
+    case BSON.CODE => TType.STRING
+    case BSON.SYMBOL => TType.STRING
+    case BSON.CODE_W_SCOPE => TType.STRING
+    case BSON.NUMBER_INT => TType.I32
+    case BSON.TIMESTAMP => TType.I64
+    case BSON.NUMBER_LONG => TType.I64
+  }
+
+  /**
+   * Reading methods.
+   */
+
+  def readMessageBegin(): TMessage = {
+    TBSONBinaryProtocol.ANONYMOUS_MESSAGE
+  }
+
+  def readMessageEnd(): Unit = {
+  }
+
+  def readStructBegin(): TStruct = {
+    if (readStack.size == 0) {
+      readStack.push(new StructReadState(inputStream, buffer))
+    } else {
+      readStack.push(currentState().readStruct())
+    }
+    TBSONBinaryProtocol.ANONYMOUS_STRUCT
+  }
+
+  def readStructEnd(): Unit = {
+    val readState = popReadState(classOf[StructReadState])
+    readState.readEnd()
+  }
+
+  def readFieldBegin(): TField = {
+    val readState = checkReadState(readStack.peek(), classOf[StructReadState])
+    if (readState.hasAnotherField) {
+      readState.readFieldType()
+      new TField(readState.lastFieldName, getTType(readState.lastFieldType), -1)
+    } else {
+      TBSONBinaryProtocol.NO_MORE_FIELDS
+    }
+  }
+
+  def readFieldEnd(): Unit = {
+  }
+
+  def readMapBegin(): TMap = {
+    val mapReadState = currentState().readMap()
+    readStack.push(mapReadState)
+    new TMap(TType.STRING, getTType(mapReadState.lastFieldType), mapReadState.itemCount)
+  }
+
+  def readMapEnd(): Unit = {
+    popReadState(classOf[MapReadState])
+  }
+
+  def readListBegin(): TList = {
+    val listReadState = currentState().readList()
+    readStack.push(listReadState)
+    new TList(getTType(listReadState.lastFieldType), listReadState.itemCount)
+  }
+
+  def readListEnd(): Unit = {
+    popReadState(classOf[ListReadState])
+  }
+
+  def readSetBegin(): TSet = {
+    val listReadState = currentState().readList()
+    readStack.push(listReadState)
+    new TSet(getTType(listReadState.lastFieldType), listReadState.itemCount)
+  }
+
+  def readSetEnd(): Unit = {
+    popReadState(classOf[ListReadState])
+  }
+
+  def readBool(): Boolean = {
+    currentState().readBool()
+  }
+
+  def readByte(): Byte = {
+    currentState().readI32().toByte
+  }
+
+  def readI16(): Short = {
+    currentState().readI32().toShort
+  }
+
+  def readI32(): Int = {
+    val readState = currentState()
+    val intValue = currentState().readI32()
+    // hack to keep track of mongo error
+    if (_errorMessage != null && readState.lastFieldName == TBSONBinaryProtocol.CODE_KEY) {
+      _errorCode = intValue
+    }
+    intValue
+  }
+
+  def readI64(): Long = {
+    val readState = currentState()
+    // be lenient here to handle legacy records written out in i32
```

Comment:
BTW now that I think of it, it wasn't the mongo driver that was doing this (it has no concept of schematic type), it was Lift...